0.1.2.0 build failure on Apple Silicon

Hi there, Homebrew maintainer here. Since GHC 8.10.5 now works on Apple Silicon, we're trying to build some GHC-dependent packages for it. We've failed to build git-annex and that seems to be from a failure to build one of its (recursive) dependencies, criterion-measurement.

This is the error log:

```
Failed to build criterion-measurement-0.1.2.0.
Build log (
/private/tmp/git-annex-20210621-35608-1xqmy6o/git-annex-8.20210621/.brew_home/.cabal/logs/ghc-8.10.5/crtrn-msrmnt-0.1.2.0-dd174088.log
):
Configuring library for criterion-measurement-0.1.2.0..
clang: warning: -Wl,-headerpad_max_install_names: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries' [-Wunused-command-line-argument]
Preprocessing library for criterion-measurement-0.1.2.0..
Building library for criterion-measurement-0.1.2.0..
[1 of 3] Compiling Criterion.Measurement.Types.Internal ( src/Criterion/Measurement/Types/Internal.hs, dist/build/Criterion/Measurement/Types/Internal.o, dist/build/Criterion/Measurement/Types/Internal.dyn_o )
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
[2 of 3] Compiling Criterion.Measurement.Types ( src/Criterion/Measurement/Types.hs, dist/build/Criterion/Measurement/Types.o, dist/build/Criterion/Measurement/Types.dyn_o )
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
[3 of 3] Compiling Criterion.Measurement ( src/Criterion/Measurement.hs, dist/build/Criterion/Measurement.o, dist/build/Criterion/Measurement.dyn_o )
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
clang: warning: -Wl,-headerpad_max_install_names: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries' [-Wunused-command-line-argument]
You are using an unsupported version of LLVM!
Currently only 10 to 12 is supported. System LLVM version: 12.0.0
We will try though...
clang: warning: -Wl,-headerpad_max_install_names: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries' [-Wunused-command-line-argument]
clang: warning: -Wl,-headerpad_max_install_names: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries' [-Wunused-command-line-argument]

cbits/cycles.c:55:2: error:
     error: Unsupported OS/architecture/compiler!
   |
55 | #error Unsupported OS/architecture/compiler!
   |  ^
#error Unsupported OS/architecture/compiler!
 ^
1 error generated.
`clang' failed in phase `C Compiler'. (Exit code: 1)
cabal: Failed to build criterion-measurement-0.1.2.0 (which is required by
git-annex-8.20210621). See the build log above for details.
```

The warning about the LLVM version appears to be [spurious](https://github.com/Homebrew/homebrew-core/pull/78821#issuecomment-856519925). It seems that the released version of criterion-measurement just doesn't recognise the attempt to build for Darwin on ARM64/AArch64.

Build logs for git-annex are available at https://github.com/Homebrew/homebrew-core/actions/runs/958143882, and the error from criterion-measurement starts at https://github.com/Homebrew/homebrew-core/runs/2878261603?check_suite_focus=true#step:6:1653.

Assistance would be appreciated.

--- a/criterion-measurement/cbits/cycles.c
+++ b/criterion-measurement/cbits/cycles.c
@@ -1,6 +1,15 @@
 #include "Rts.h"
 
-#if x86_64_HOST_ARCH || i386_HOST_ARCH
+#if darwin_HOST_OS
+
+#include <mach/mach_time.h>
+
+StgWord64 criterion_rdtsc(void)
+{
+  return mach_absolute_time();
+}
+
+#elif x86_64_HOST_ARCH || i386_HOST_ARCH
 
 StgWord64 criterion_rdtsc(void)
 {

--- a/criterion-measurement/cbits/time-osx.c
+++ b/criterion-measurement/cbits/time-osx.c
@@ -1,20 +1,11 @@
 #include <mach/mach.h>
-#include <mach/mach_time.h>
+#include <time.h>
 
-static mach_timebase_info_data_t timebase_info;
-static double timebase_recip;
-
-void criterion_inittime(void)
-{
-    if (timebase_recip == 0) {
-	mach_timebase_info(&timebase_info);
-	timebase_recip = (timebase_info.denom / timebase_info.numer) / 1e9;
-    }
-}
+void criterion_inittime(void) {}
 
 double criterion_gettime(void)
 {
-    return mach_absolute_time() * timebase_recip;
+    return clock_gettime_nsec_np(CLOCK_UPTIME_RAW) / 1e9;
 }
 
 static double to_double(time_value_t time)

--- a/criterion-measurement/changelog.md
+++ b/criterion-measurement/changelog.md
@@ -1,3 +1,28 @@
+next
+
+* Change `criterion_rdtsc` to return `mach_absolute_time` on macOS. This is a
+  portable way of returning the number of CPU cycles that works on both Intel-
+  and ARM-based Macs.
+
+* Change `criterion_gettime` to use `clock_gettime_nsec_np` instead of
+  `mach_absolute_time` on macOS. While `mach_absolute_time` has nanosecond
+  resolution on Intel-based Macs, this is not the case on ARM-based Macs, so
+  the previous `mach_absolute_time`-based implementation would return incorrect
+  timing results on Apple silicon.
+
+  There are two minor consequences of this change:
+
+  * `criterion-measurement` now only supports macOS 10.02 or later, as that is
+    the first version to have `clock_gettime_nsec_np`. As macOS 10.02 was
+    released in 2002, this is unlikely to affect users, but please speak up if
+    this is a problem for you.
+
+  * As `clock_gettime_nsec_np` does not require any special initialization
+    code, `criterion_inittime` is now a no-op on macOS. If you manually invoke
+    the `getTime` function in your code, however, it is still important that
+    you `initializeTime` beforehand, as this is still required for the Windows
+    implementation to work correctly.
+
 0.1.2.0
 
 * Ensure that `Criterion.Measurement.Types.Internal` is always compiled with

